### ISSUE TITLE
Method to read a set of irregular files

### ISSUE THREAD

**Issue by neishm** [pygeode/pygeode#95]
As an aside for #48, there may be perfectly valid files that simply can't be opened with `open_multi`.  For instance, I've run in the the following situations when running our model (GEM):
- Partway through an experiment, we may realize we forgot to output a particular field.  Rather than start from scratch, we tweak the output config and continue on.  End result is we're missing a field for part of a model integration.
- Conversely, we may decide that we don't really need a certain field, and stop outputting it partway through a model run.
- Due to space constraints, we usually limit the full 3D output to only once every 24h, and output the surface level every 2h.  So, every n<sup>th</sup> file will have a larger vertical axis than the rest of the files.
- Some fields may be output at different intervals from the rest (e.g., maybe we output surface CO<sub>2</sub> every 2 hours, but maybe another tracer every 4 hours).
- For older experiments, the model output is split into multiple files per timestep.  E.g., for a 24h forecast from `2009-01-01 00:00` we may have a file named `dm2009010100_024` with "dynamics" output, and `pm2009010100_024` with "physics" output.  Both files have the same timestep, but mutually exclusive fields.

End result is, `open_multi` tells us to f off.

To deal with these files, I've constructed a utility method in our (EC) diagnostic package which does the following:
1. Opens every single file, one at a time, and inspects the contents (field names and axes).  This information is saved into a cache file, so if we use the same files in the future we can skip this step.
2. Finds every possible way to organize this data into regular datasets.  For instance, it determines that it can construct a dataset with surface-level output every 2 hours, and it can also construct a dataset with 3D output every 24 hours, etc.  There may be many such datasets with various combinations of fields / axes.  This is done as general as possible, so for instance if you want to split the data into northern / southern hemisphere files, or something even more annoying, it should figure out how to stitch everything back together.
3. Stores all these datasets in a container class, and provides some methods for querying the data to find the best dataset for a particular purpose.

For example, if I want to get the full 3D fields for CO<sub>2</sub>:

``` python
f = DataInterface.from_files(glob("mydata/*"), opener=fstd.open, manifest="mydata.save")
full_levels = lambda d: len(d.levels) == 80
co2 = f.find_best("CO2", requirement=full_levels)
```

To get CO<sub>2</sub> and CH<sub>4</sub> co-located at the same points, at the highest output frequency (don't care about which levels):

``` python
number_of_timesteps = lambda d: len(d.time)
co2, ch4 = f.find_best(["CO2","CH4"], maximize=number_of_timesteps)
```

The functionality is currently embedded within my diagnostic package, but I'm wondering if it might be worthwhile to factor it out, and bring it into PyGeode?  It may be useful as a last resort when `open_multi` and `openall` fail.

**Comment by neishm** [comment]
Just pushed a version of this module to a `data_scanner` feature branch (had to wait for the code to get released under an open-source license).

An updated pseudo-example of how to use it:
```python
from pygeode.formats.data_scanner import DataInterface, from_files
f = DataInterface(from_files("/path/to/your/data/xyz*.nc", interface="netcdf", manifest="xyz.save"))
# Find CO2 that has full vertical extent (e.g. ignore any high-frequency surface output in the files).
number_of_levels = lambda dataset: len(dataset.zaxis)
co2 = f.find_best("CO2", maximize=number_of_levels)
# Similar, but find both CO2 and CH4 constrained to be co-located at the same lat/lon/height/time
# (in case, for instance, 3D CH4 is available less frequently than CO2)
co2, ch4 = f.find_best(["CO2","CH4"], maximize=number_of_levels)
```
I'll be leaving this in the `data_scanner` feature branch for a while, since this is not part of the upcoming v1.1 release.  As such, it's still very much open to suggestions for improvements or other changes.

**Comment by neishm** [comment]
Merged this into the master branch, should be available in release-1.1.

**Comment by neishm** [comment]
Re-opening, so I remember to document this!
